### ISSUE TITLE
[Java] CloneMethodMustImplementCloneable doesn't check inherited implementation of Cloneable

### ISSUE THREAD

**Issue by demonfiddler** [pmd/pmd#574]
**Rule Set:** Type Resolution

**Description:** The CloneMethodMustImplementCloneable rule can erroneously report a violation because it doesn't bother to check whether the class in question extends one that _does_ implement Cloneable.

**Code Sample demonstrating the issue:**

```
class MyArrayList<T> extends ArrayList<T> {
	// PMD.CloneMethodMustImplementCloneable incorrectly flags this method as a violation.
	@Override
	public Object clone() {
		throw new UnsupportedOperationException();
	}
}
```

**Running PMD through:** *[CLI | Eclipse | Maven]*

**Comment by jsotuyod** [comment]
@demonfiddler thanks for the report! Just to be sure, but are you certain you are using the rule from [`java-typeresolution/CloneMethodMustImplementCloneable`](https://pmd.github.io/pmd-5.8.1/pmd-java/rules/java/typeresolution.html#CloneMethodMustImplementCloneable) and not the one from [`java-clone/CloneMethodMustImplementCloneable`](https://pmd.github.io/pmd-5.8.1/pmd-java/rules/java/clone.html#CloneMethodMustImplementCloneable)?

The rules are called the same, but them being different rulesets mean they have different implementations. The one from the `clone` ruleset will for certain reproduce this issue, the one from `typeresolution` ruleset shouldn't (as long as the auxclasspath is properly filled).

If you are using the one from `typeresolution`, are you certain you are passing the auxclasspath info? In Maven this is done through the [`typeResolution` configuration attribute](https://maven.apache.org/plugins/maven-pmd-plugin/pmd-mojo.html#typeResolution) (defaults to true in the plugin 3.8, but used to be false). For CLI you need to set it manually through the [`-auxclasspath` parameter](https://pmd.github.io/pmd-5.8.1/usage/running.html).

For Eclipse I'm not certain it is properly populated by default, it should if it's not. @adangel can probably tell.

**Comment by demonfiddler** [comment]
@jsotuyod thanks for the detailed response. Yes, this was definitely occurring with the version in the typeresolution ruleset. I believe I was seeing the same problem when invoking PMD from both Eclipse and Maven but for reasons I've not yet fathomed the Eclipse PMD builder seems to stopped reporting any PMD issues on this particular project and right now I'm no longer seeing it from either. I hope I haven't raised a false alarm. Will post again when I've done a deeper investigation.

**Comment by jsotuyod** [comment]
@demonfiddler I'm just noticing the updated comment. Good to hear.

This is possibly an Eclipse plugin issue, not providing a complete auxclasspath. Just to be sure 'though, which PMD plugin are you using? As far as I know there are 2 plugins:

1. https://github.com/acanda/eclipse-pmd
2. https://github.com/pmd/pmd-eclipse-plugin

Thanks in advanced

**Comment by demonfiddler** [comment]
@jsotuyod I'm using #2, pmd-eclipse-plugin. (Fairly sure about that, will check tomorrow when I'm back in the office.)
